File: 0496-next-greater-element-i/0496-next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        unordered_map<int,int> mp;
        stack<int> st;
        vector<int> res(nums2.size(),-1);
        
        for(int i=nums2.size()-1;i>=0;i--){
            while(st.size()>0 and st.top()<=nums2[i]){
                st.pop();
            }    
            if(st.size()>0)
                res[i] = st.top();
            
            st.push(nums2[i]);
        }
        
        for(int i=0;i<nums2.size();i++){
            mp[nums2[i]] = res[i];
        }
        
        vector<int> result;
        
        for(auto &x:nums1){
            result.emplace_back(mp[x]);
        }
        
        return result;
    }
};

File: 0496-next-greater-element-i/0496-next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        mp = dict()
        st = []
        res = [-1]*len(nums2)
        
        for i in range(len(nums2)-1,-1,-1):
            if not st:
                res[i]=-1
            
            while st and not st[-1] > nums2[i]:
                st.pop()
                
            if st:
                res[i] = st[-1]
            
            st.append(nums2[i])
            
        for i,ele in enumerate(nums2):
            mp[ele] = res[i]
            
        return [mp[ele] for ele in nums1]
            